File: harm_osc/bootstrap.py
import pandas as pd
import numpy as np
import sys

input_file = sys.argv[1]

# Read the dataset into a Pandas DataFrame
df = pd.read_csv(input_file, names=["Time", "Value"])

# Group values by time
grouped_data = df.groupby("Time")["Value"]
print(grouped_data)
# Define the number of bootstraps
num_bootstraps = 100

# Perform bootstrapping
bootstrap_means = []

for _ in range(num_bootstraps):
    bootstrap_sample = []

    # Resample values while keeping the same time
    for time, values in grouped_data:
        resampled_values = np.random.choice(values, len(values), replace=True)
        bootstrap_sample.extend(resampled_values)

    # Calculate the mean of the bootstrap sample
    bootstrap_mean = np.mean(bootstrap_sample)
    bootstrap_means.append(bootstrap_mean)


if len(sys.argv) > 2:
    output_file = sys.argv[2]
else:
    output_file = "bootstrap_results.dat"


with open(output_file, "w") as file:
    file.write(f"Time \t \t Mean_Bootstrap\n")
    for i, mean in enumerate(bootstrap_means):
        file.write(f"{i+1} \t \t {mean:.9f}\n")

print(f"Bootstrap results have been written to {output_file}")

File: harm_osc/bootstrap.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <random>

// Structure to represent a data point
struct DataPoint {
    int time;
    double value;
};

int main(int argc, char *argv[]) {
	char in_string[50], out_string[50];
	while( argc > 1 ) {

            switch(argv[1][0]) {
            case '-':
				if(argv[1][1] == 'i') std::sprintf(in_string, "%s", &argv[1][2]);
				else if(argv[1][1] == 'o') std::sprintf(out_string, "%s", &argv[1][2]);
                break;
        	}
        	++argv;
        	--argc;
	}

    // Define the number of bootstraps
    int numBootstraps = 1000;

    // Read the dataset from the file into memory
    // std::cout << in_string << std::endl;
    std::ifstream inputFile(in_string);
    if (!inputFile.is_open()) {
        std::cerr << "Error opening file." << std::endl;
        return 1;
    }

    std::vector<DataPoint> dataset;
    int time;
    double value;
    while (inputFile >> time >> value) {
        dataset.push_back({time, value});
    }
    inputFile.close();

    // Create a map to group values by time
    std::map<int, std::vector<double>> groupedData;
    for (const auto& dataPoint : dataset) {
        groupedData[dataPoint.time].push_back(dataPoint.value);
    }

    // Set up a random number generator for sampling with replacement
    std::random_device rd;
    std::mt19937 gen(rd());

    std::ofstream outputFile(out_string);
    if (!outputFile.is_open()) {
        std::cerr << "Error opening output file." << std::endl;
        return 1;
    }

    outputFile << "# BootstrapNum \t \t time \t\t Corr" << std::endl;
    // Perform bootstrapping
    for (const auto& pair : groupedData) {
            int currentTime = pair.first;
            const std::vector<double>& values = pair.second;

            for (int bootstrap = 0; bootstrap < numBootstraps; ++bootstrap) {
                std::vector<double> bootstrapSample;

                // Resample values for the current time
                for (int i = 0; i < int(values.size()); ++i) {
                    int randomIndex = std::uniform_int_distribution<int>(0, values.size() - 1)(gen);
                    bootstrapSample.push_back(values[randomIndex]);
                }

                // Calculate the mean of the bootstrap sample
                double mean = 0.0;
                for (double val : bootstrapSample) {
                    mean += val;
                }
                mean /= bootstrapSample.size();

                //Print or store the results for each bootstrap
                outputFile << bootstrap + 1 << " \t \t " << currentTime << " \t \t " << mean << std::endl;
            }
    }

    return 0;
}

